File: hkws_Development/Samples/C#/SimpleSamples/ColorCorrect/ColorCorrect.cs
using System;
using System.Collections.Generic;
using MvCamCtrl.NET;
using System.Runtime.InteropServices;
using System.IO;

namespace ColorCorrect
{
    class ColorCorrect
    {
        static void Main(string[] args)
        {
            int nRet = MyCamera.MV_OK;
            MyCamera device = new MyCamera();
            IntPtr pDstData = IntPtr.Zero;
            UInt32 nDstDataSize = 0;

            IntPtr pCLUTData = IntPtr.Zero;
            UInt32 nCLUTDataSize = 0;

            do
            {
                // ch:枚举设备 | en:Enum deivce
                MyCamera.MV_CC_DEVICE_INFO_LIST stDevList = new MyCamera.MV_CC_DEVICE_INFO_LIST();
                nRet = MyCamera.MV_CC_EnumDevices_NET(MyCamera.MV_GIGE_DEVICE | MyCamera.MV_USB_DEVICE, ref stDevList);
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Enum device failed:{0:x8}", nRet);
                    break;
                }
                Console.WriteLine("Enum device count :{0} \n", stDevList.nDeviceNum);
                if (0 == stDevList.nDeviceNum)
                {
                    break;
                }

                MyCamera.MV_CC_DEVICE_INFO stDevInfo;

                // ch:打印设备信息 | en:Print device info
                for (Int32 i = 0; i < stDevList.nDeviceNum; i++)
                {
                    stDevInfo = (MyCamera.MV_CC_DEVICE_INFO)Marshal.PtrToStructure(stDevList.pDeviceInfo[i], typeof(MyCamera.MV_CC_DEVICE_INFO));

                    if (MyCamera.MV_GIGE_DEVICE == stDevInfo.nTLayerType)
                    {
                        MyCamera.MV_GIGE_DEVICE_INFO stGigEDeviceInfo = (MyCamera.MV_GIGE_DEVICE_INFO)MyCamera.ByteToStruct(stDevInfo.SpecialInfo.stGigEInfo, typeof(MyCamera.MV_GIGE_DEVICE_INFO));
                        uint nIp1 = ((stGigEDeviceInfo.nCurrentIp & 0xff000000) >> 24);
                        uint nIp2 = ((stGigEDeviceInfo.nCurrentIp & 0x00ff0000) >> 16);
                        uint nIp3 = ((stGigEDeviceInfo.nCurrentIp & 0x0000ff00) >> 8);
                        uint nIp4 = (stGigEDeviceInfo.nCurrentIp & 0x000000ff);
                        Console.WriteLine("[device " + i.ToString() + "]:");
                        Console.WriteLine("DevIP:" + nIp1 + "." + nIp2 + "." + nIp3 + "." + nIp4);
                        Console.WriteLine("UserDefineName:" + stGigEDeviceInfo.chUserDefinedName + "\n");
                    }
                    else if (MyCamera.MV_USB_DEVICE == stDevInfo.nTLayerType)
                    {
                        MyCamera.MV_USB3_DEVICE_INFO stUsb3DeviceInfo = (MyCamera.MV_USB3_DEVICE_INFO)MyCamera.ByteToStruct(stDevInfo.SpecialInfo.stUsb3VInfo, typeof(MyCamera.MV_USB3_DEVICE_INFO));
                        Console.WriteLine("[device " + i.ToString() + "]:");
                        Console.WriteLine("SerialNumber:" + stUsb3DeviceInfo.chSerialNumber);
                        Console.WriteLine("UserDefineName:" + stUsb3DeviceInfo.chUserDefinedName + "\n");
                    }
                }

                Int32 nDevIndex = 0;
                Console.Write("Please input index(0-{0:d}):", stDevList.nDeviceNum - 1);
                try
                {
                    nDevIndex = Convert.ToInt32(Console.ReadLine());
                }
                catch
                {
                    Console.Write("Invalid Input!\n");
                    break;
                }

                if (nDevIndex > stDevList.nDeviceNum - 1 || nDevIndex < 0)
                {
                    Console.Write("Input Error!\n");
                    break;
                }
                stDevInfo = (MyCamera.MV_CC_DEVICE_INFO)Marshal.PtrToStructure(stDevList.pDeviceInfo[nDevIndex], typeof(MyCamera.MV_CC_DEVICE_INFO));

                // ch:创建设备 | en: Create device
                nRet = device.MV_CC_CreateDevice_NET(ref stDevInfo);
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Create device failed:{0:x8}", nRet);
                    break;
                }

                // ch:打开设备 | en:Open device
                nRet = device.MV_CC_OpenDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Open device failed:{0:x8}", nRet);
                    break;
                }

                // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
                if (stDevInfo.nTLayerType == MyCamera.MV_GIGE_DEVICE)
                {
                    int nPacketSize = device.MV_CC_GetOptimalPacketSize_NET();
                    if (nPacketSize > 0)
                    {
                        nRet = device.MV_CC_SetIntValue_NET("GevSCPSPacketSize", (uint)nPacketSize);
                        if (nRet != MyCamera.MV_OK)
                        {
                            Console.WriteLine("Warning: Set Packet Size failed {0:x8}", nRet);
                        }
                    }
                    else
                    {
                        Console.WriteLine("Warning: Get Packet Size failed {0:x8}", nPacketSize);
                    }
                }

                // ch:设置触发模式为off || en:set trigger mode as off
                nRet = device.MV_CC_SetEnumValue_NET("TriggerMode", 0);
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Set TriggerMode failed:{0:x8}", nRet);
                    break;
                }

                // ch:开启抓图 || en: start grab image
                nRet = device.MV_CC_StartGrabbing_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Start grabbing failed:{0:x8}", nRet);
                    break;
                }

                MyCamera.MV_FRAME_OUT stFrameOut = new MyCamera.MV_FRAME_OUT();

                nRet = device.MV_CC_GetImageBuffer_NET(ref stFrameOut, 1000);
                // 获取一帧图像
                if (MyCamera.MV_OK == nRet)
                {
                    Console.WriteLine("Get One Frame:" + "Width[" + Convert.ToString(stFrameOut.stFrameInfo.nWidth) + "] , Height["
                        + Convert.ToString(stFrameOut.stFrameInfo.nHeight) + "] , FrameNum[" + Convert.ToString(stFrameOut.stFrameInfo.nFrameNum) + "]");

                    if (pDstData == IntPtr.Zero || nDstDataSize < stFrameOut.stFrameInfo.nFrameLen)
                    {
                        if (pDstData != IntPtr.Zero)
                        {
                            Marshal.FreeHGlobal(pDstData);
                            pDstData = IntPtr.Zero;
                            nDstDataSize = 0;
                        }

                        pDstData = Marshal.AllocHGlobal((int)stFrameOut.stFrameInfo.nFrameLen);
                        if (pDstData == IntPtr.Zero)
                        {
                            Console.WriteLine("malloc pDstData failed");
                            nRet = MyCamera.MV_E_RESOURCE;
                            break;
                        }
                        nDstDataSize = stFrameOut.stFrameInfo.nFrameLen;
                    }

                    MyCamera.MV_CC_COLOR_CORRECT_PARAM stColorCorrectParam = new MyCamera.MV_CC_COLOR_CORRECT_PARAM();

                    //图像相关
                    stColorCorrectParam.nWidth = stFrameOut.stFrameInfo.nWidth;
                    stColorCorrectParam.nHeight = stFrameOut.stFrameInfo.nHeight;
                    stColorCorrectParam.pSrcBuf = stFrameOut.pBufAddr;
                    stColorCorrectParam.nSrcBufLen = stFrameOut.stFrameInfo.nFrameLen;
                    stColorCorrectParam.enPixelType = stFrameOut.stFrameInfo.enPixelType;

                    stColorCorrectParam.pDstBuf = pDstData;
                    stColorCorrectParam.nDstBufSize = nDstDataSize;
                    stColorCorrectParam.nImageBit = 8;

                    //Gamma参数，可根据实际情况选择类型
                    stColorCorrectParam.stGammaParam.enGammaType = MyCamera.MV_CC_GAMMA_TYPE.MV_CC_GAMMA_TYPE_VALUE;
                    stColorCorrectParam.stGammaParam.fGammaValue = (float)0.6;

                    //CCM参数和CLUT参数至少输入一种，也可以两种一起处理
                    stColorCorrectParam.stCCMParam.bCCMEnable = false;

                    stColorCorrectParam.stCLUTParam.bCLUTEnable = true;
                    stColorCorrectParam.stCLUTParam.nCLUTScale = 1024;
                    stColorCorrectParam.stCLUTParam.nCLUTSize = 17;

                    if (false == File.Exists("./CLUTCalib.bin"))
                    {
                        Console.WriteLine("Open file failed");
                        nRet = MyCamera.MV_E_RESOURCE;
                        break;
                    }

                    FileStream fs = new FileStream("./CLUTCalib.bin", FileMode.Open);
                    byte[] data = new byte[fs.Length];

                    if (pCLUTData == IntPtr.Zero || nCLUTDataSize < fs.Length)
                    {
                        if (pCLUTData != IntPtr.Zero)
                        {
                            Marshal.FreeHGlobal(pCLUTData);
                            pCLUTData = IntPtr.Zero;
                            nCLUTDataSize = 0;
                        }

                        pCLUTData = Marshal.AllocHGlobal((int)fs.Length);
                        if (pCLUTData == IntPtr.Zero)
                        {
                            Console.WriteLine("malloc pCLUTData failed");
                            break;
                        }
                        nCLUTDataSize = (uint)fs.Length;
                    }

                    fs.Read(data, 0, data.Length);
                    fs.Close();

                    Marshal.Copy(data, 0, pCLUTData, (Int32)nCLUTDataSize);

                    stColorCorrectParam.stCLUTParam.pCLUTBuf = pCLUTData;
                    stColorCorrectParam.stCLUTParam.nCLUTBufLen = nCLUTDataSize;

                    nRet = device.MV_CC_ColorCorrect_NET(ref stColorCorrectParam);
                    if (MyCamera.MV_OK != nRet)
                    {
                        Console.WriteLine("Color Correct Failed:{0:x8}", nRet);
                        break;
                    }

                    //保存图像到文件
                    MyCamera.MV_SAVE_IMG_TO_FILE_PARAM stSaveFileParam = new MyCamera.MV_SAVE_IMG_TO_FILE_PARAM();
                    stSaveFileParam.enImageType = MyCamera.MV_SAVE_IAMGE_TYPE.MV_Image_Bmp;
                    stSaveFileParam.enPixelType = stFrameOut.stFrameInfo.enPixelType;
                    stSaveFileParam.nWidth = stFrameOut.stFrameInfo.nWidth;
                    stSaveFileParam.nHeight = stFrameOut.stFrameInfo.nHeight;
                    stSaveFileParam.nDataLen = stFrameOut.stFrameInfo.nFrameLen;
                    stSaveFileParam.pData = stFrameOut.pBufAddr;
                    stSaveFileParam.pImagePath = "BeforeImage_w" + stSaveFileParam.nWidth.ToString() + "_" + stSaveFileParam.nHeight.ToString() + "_fn" + stFrameOut.stFrameInfo.nFrameNum.ToString() + ".bmp";
                    nRet = device.MV_CC_SaveImageToFile_NET(ref stSaveFileParam);
                    if (MyCamera.MV_OK != nRet)
                    {
                        Console.WriteLine("SaveImageToFile failed:{0:x8}", nRet);
                        break;
                    }

                    stSaveFileParam.pData = pDstData;
                    stSaveFileParam.pImagePath = "AfterImage_w" + stSaveFileParam.nWidth.ToString() + "_" + stSaveFileParam.nHeight.ToString() + "_fn" + stFrameOut.stFrameInfo.nFrameNum.ToString() + ".bmp";
                    nRet = device.MV_CC_SaveImageToFile_NET(ref stSaveFileParam);
                    if (MyCamera.MV_OK != nRet)
                    {
                        Console.WriteLine("SaveImageToFile failed:{0:x8}", nRet);
                        break;
                    }

                    device.MV_CC_FreeImageBuffer_NET(ref stFrameOut);
                }
                else
                {
                    Console.WriteLine("Get Image failed:{0:x8}", nRet);
                }

                // ch:停止抓图 | en:Stop grabbing
                nRet = device.MV_CC_StopGrabbing_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Stop grabbing failed:{0:x8}", nRet);
                    break;
                }

                // ch:关闭设备 | en:Close device
                nRet = device.MV_CC_CloseDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Close device failed:{0:x8}", nRet);
                    break;
                }

                // ch:销毁设备 | en:Destroy device
                nRet = device.MV_CC_DestroyDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Destroy device failed:{0:x8}", nRet);
                    break;
                }
            } while (false);

            if (pDstData != IntPtr.Zero)
            {
                Marshal.FreeHGlobal(pDstData);
                pDstData = IntPtr.Zero;
            }

            if (pCLUTData != IntPtr.Zero)
            {
                Marshal.FreeHGlobal(pCLUTData);
                pCLUTData = IntPtr.Zero;
            }

            if (MyCamera.MV_OK != nRet)
            {
                // ch:销毁设备 | en:Destroy device
                nRet = device.MV_CC_DestroyDevice_NET();
                if (MyCamera.MV_OK != nRet)
                {
                    Console.WriteLine("Destroy device failed:{0:x8}", nRet);
                }
            }

            Console.WriteLine("Press enter to exit");
            Console.ReadKey();
        }
    }
}

File: hkws_Development/Samples/VC/VS/SimpleSamples/ColorCorrect/ColorCorrect.cpp
#include <stdio.h>
#include <Windows.h>
#include <conio.h>
#include <io.h>
#include "MvCameraControl.h"

// ch:等待按键输入 | en:Wait for key press
void WaitForKeyPress(void)
{
    while(!_kbhit())
    {
        Sleep(10);
    }
    _getch();
}

bool PrintDeviceInfo(MV_CC_DEVICE_INFO* pstMVDevInfo)
{
    if (NULL == pstMVDevInfo)
    {
        printf("The Pointer of pstMVDevInfo is NULL!\n");
        return false;
    }
    if (pstMVDevInfo->nTLayerType == MV_GIGE_DEVICE)
    {
        int nIp1 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0xff000000) >> 24);
        int nIp2 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x00ff0000) >> 16);
        int nIp3 = ((pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x0000ff00) >> 8);
        int nIp4 = (pstMVDevInfo->SpecialInfo.stGigEInfo.nCurrentIp & 0x000000ff);

        // ch:打印当前相机ip和用户自定义名字 | en:print current ip and user defined name
        printf("CurrentIp: %d.%d.%d.%d\n" , nIp1, nIp2, nIp3, nIp4);
        printf("UserDefinedName: %s\n\n" , pstMVDevInfo->SpecialInfo.stGigEInfo.chUserDefinedName);
    }
    else if (pstMVDevInfo->nTLayerType == MV_USB_DEVICE)
    {
        printf("UserDefinedName: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chUserDefinedName);
        printf("Serial Number: %s\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.chSerialNumber);
        printf("Device Number: %d\n\n", pstMVDevInfo->SpecialInfo.stUsb3VInfo.nDeviceNumber);
    }
    else
    {
        printf("Not support.\n");
    }

    return true;
}

int main()
{
    int nRet = MV_OK;
    void* handle = NULL;
    unsigned char *pDstData = NULL;
    unsigned int nDstDataSize = 0;

    unsigned char *pCLUTData = NULL;
    unsigned int nCLUTDataSize = 0;

    do 
    {
        // ch:枚举设备 | en:Enum device
        MV_CC_DEVICE_INFO_LIST stDeviceList;
        memset(&stDeviceList, 0, sizeof(MV_CC_DEVICE_INFO_LIST));
        nRet = MV_CC_EnumDevices(MV_GIGE_DEVICE | MV_USB_DEVICE, &stDeviceList);
        if (MV_OK != nRet)
        {
            printf("Enum Devices fail! nRet [0x%x]\n", nRet);
            break;
        }

        if (stDeviceList.nDeviceNum > 0)
        {
            for (unsigned int i = 0; i < stDeviceList.nDeviceNum; i++)
            {
                printf("[device %d]:\n", i);
                MV_CC_DEVICE_INFO* pDeviceInfo = stDeviceList.pDeviceInfo[i];
                if (NULL == pDeviceInfo)
                {
                    break;
                } 
                PrintDeviceInfo(pDeviceInfo);            
            }  
        } 
        else
        {
            printf("Find No Devices!\n");
            break;
        }

        printf("Please Input camera index(0-%d):", stDeviceList.nDeviceNum-1);
        unsigned int nIndex = 0;
        scanf_s("%d", &nIndex);

        if (nIndex >= stDeviceList.nDeviceNum)
        {
            printf("Input error!\n");
            break;
        }

        // ch:选择设备并创建句柄 | en:Select device and create handle
        nRet = MV_CC_CreateHandle(&handle, stDeviceList.pDeviceInfo[nIndex]);
        if (MV_OK != nRet)
        {
            printf("Create Handle fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:打开设备 | en:Open device
        nRet = MV_CC_OpenDevice(handle);
        if (MV_OK != nRet)
        {
            printf("Open Device fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:探测网络最佳包大小(只对GigE相机有效) | en:Detection network optimal package size(It only works for the GigE camera)
        if (stDeviceList.pDeviceInfo[nIndex]->nTLayerType == MV_GIGE_DEVICE)
        {
            int nPacketSize = MV_CC_GetOptimalPacketSize(handle);
            if (nPacketSize > 0)
            {
                nRet = MV_CC_SetIntValue(handle,"GevSCPSPacketSize",nPacketSize);
                if(nRet != MV_OK)
                {
                    printf("Warning: Set Packet Size fail nRet [0x%x]!", nRet);
                }
            }
            else
            {
                printf("Warning: Get Packet Size fail nRet [0x%x]!", nPacketSize);
            }
        }

        nRet = MV_CC_SetEnumValue(handle, "TriggerMode", MV_TRIGGER_MODE_OFF);
        if (MV_OK != nRet)
        {
            printf("Set Trigger Mode fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:开始取流 | en:Start grab image
        nRet = MV_CC_StartGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Start Grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }

        MV_FRAME_OUT stImageInfo = {0};

        nRet = MV_CC_GetImageBuffer(handle, &stImageInfo, 1000);
        if (nRet == MV_OK)
        {
            printf("Get One Frame: Width[%d], Height[%d], nFrameNum[%d]\n", 
                stImageInfo.stFrameInfo.nWidth, stImageInfo.stFrameInfo.nHeight, stImageInfo.stFrameInfo.nFrameNum);

            if (pDstData == NULL || nDstDataSize < stImageInfo.stFrameInfo.nFrameLen)
            {
                if (pDstData)
                {
                    free(pDstData);
                    pDstData = NULL;
                    nDstDataSize = 0;
                }

                pDstData = (unsigned char *)malloc(stImageInfo.stFrameInfo.nFrameLen);
                if (pDstData == NULL)
                {
                    printf("malloc pDstData fail!\n");
                    nRet = MV_E_RESOURCE;
                    break;
                }
                nDstDataSize = stImageInfo.stFrameInfo.nFrameLen;
            }

            MV_CC_COLOR_CORRECT_PARAM stColorCorrectParam = {0};

            //图像相关
            stColorCorrectParam.nWidth = stImageInfo.stFrameInfo.nWidth;
            stColorCorrectParam.nHeight = stImageInfo.stFrameInfo.nHeight;
            stColorCorrectParam.pSrcBuf = stImageInfo.pBufAddr;
            stColorCorrectParam.nSrcBufLen = stImageInfo.stFrameInfo.nFrameLen;
            stColorCorrectParam.enPixelType = stImageInfo.stFrameInfo.enPixelType;

            stColorCorrectParam.pDstBuf = pDstData;
            stColorCorrectParam.nDstBufSize = nDstDataSize;
            stColorCorrectParam.nImageBit = 8;

            //Gamma参数，可根据实际情况选择类型
            stColorCorrectParam.stGammaParam.enGammaType = MV_CC_GAMMA_TYPE_VALUE;
            stColorCorrectParam.stGammaParam.fGammaValue = 0.6;

            //CCM参数和CLUT参数至少输入一种，也可以两种一起处理
            stColorCorrectParam.stCCMParam.bCCMEnable = false;

            stColorCorrectParam.stCLUTParam.bCLUTEnable = true;
            stColorCorrectParam.stCLUTParam.nCLUTScale = 1024;
            stColorCorrectParam.stCLUTParam.nCLUTSize = 17;

            FILE* fp = fopen("./CLUTCalib.bin", "rb+");
            if (fp == NULL)
            {
                printf("Failed to open 'CLUTCalib.bin', This file is required to do CLUT!\n");
                nRet = MV_E_RESOURCE;
                break;
            }

            unsigned int nNeedCLUTBufLen = stColorCorrectParam.stCLUTParam.nCLUTSize * stColorCorrectParam.stCLUTParam.nCLUTSize * stColorCorrectParam.stCLUTParam.nCLUTSize * sizeof(int) * 3;
            int nFileLen = filelength(fileno(fp));
            if (nFileLen < nNeedCLUTBufLen)
            {
                printf("The file (CLUTCalib.bin) is too small!\n");
                nRet = MV_E_BUFOVER;
                break;
            }

            if (pCLUTData == NULL || nCLUTDataSize < nNeedCLUTBufLen)
            {
                if (pCLUTData)
                {
                    free(pCLUTData);
                    pCLUTData = NULL;
                    nCLUTDataSize = 0;
                }

                pCLUTData = (unsigned char *)malloc(nNeedCLUTBufLen);
                if (pCLUTData == NULL)
                {
                    printf("malloc pCLUTData fail !\n");
                    nRet = MV_E_RESOURCE;
                    break;
                }
                nCLUTDataSize = nNeedCLUTBufLen;
            }
            fread(pCLUTData, 1, nCLUTDataSize, fp);
            fclose(fp);

            stColorCorrectParam.stCLUTParam.pCLUTBuf = pCLUTData;
            stColorCorrectParam.stCLUTParam.nCLUTBufLen = nCLUTDataSize;

            nRet = MV_CC_ColorCorrect(handle, &stColorCorrectParam);
            if (MV_OK != nRet)
            {
                printf("Color Correct fail! nRet [0x%x]\n", nRet);
                break;
            }

            //保存图像到文件
            MV_SAVE_IMG_TO_FILE_PARAM stSaveFileParam;
            memset(&stSaveFileParam, 0, sizeof(MV_SAVE_IMG_TO_FILE_PARAM));

            stSaveFileParam.enImageType = MV_Image_Bmp;
            stSaveFileParam.enPixelType = stImageInfo.stFrameInfo.enPixelType;
            stSaveFileParam.nWidth      = stImageInfo.stFrameInfo.nWidth;
            stSaveFileParam.nHeight     = stImageInfo.stFrameInfo.nHeight;
            stSaveFileParam.nDataLen    = stImageInfo.stFrameInfo.nFrameLen;
            stSaveFileParam.pData       = stImageInfo.pBufAddr;
            sprintf_s(stSaveFileParam.pImagePath, 256, "BeforeImage_w%d_h%d_fn%03d.bmp", stSaveFileParam.nWidth, stSaveFileParam.nHeight, stImageInfo.stFrameInfo.nFrameNum);
            nRet = MV_CC_SaveImageToFile(handle, &stSaveFileParam);
            if (MV_OK != nRet)
            {
                printf("SaveImageToFile failed[%x]!\n", nRet);
                break;
            }

            stSaveFileParam.pData       = pDstData;
            sprintf_s(stSaveFileParam.pImagePath, 256, "AfterImage_w%d_h%d_fn%03d.bmp", stSaveFileParam.nWidth, stSaveFileParam.nHeight, stImageInfo.stFrameInfo.nFrameNum);
            nRet = MV_CC_SaveImageToFile(handle, &stSaveFileParam);
            if (MV_OK != nRet)
            {
                printf("SaveImageToFile failed[%x]!\n", nRet);
                break;
            }

            MV_CC_FreeImageBuffer(handle, &stImageInfo);
        }
        else
        {
            printf("Get Image fail! nRet [0x%x]\n", nRet);
        }

        // ch:停止取流 | en:Stop grab image
        nRet = MV_CC_StopGrabbing(handle);
        if (MV_OK != nRet)
        {
            printf("Stop Grabbing fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:关闭设备 | en:Close device
        nRet = MV_CC_CloseDevice(handle);
        if (MV_OK != nRet)
        {
            printf("Close Device fail! nRet [0x%x]\n", nRet);
            break;
        }

        // ch:销毁句柄 | en:Destroy handle
        nRet = MV_CC_DestroyHandle(handle);
        if (MV_OK != nRet)
        {
            printf("Destroy Handle fail! nRet [0x%x]\n", nRet);
            break;
        }
    } while (0);

    if (pDstData)
    {
        free(pDstData);
        pDstData = NULL;
    }

    if (pCLUTData)
    {
        free(pCLUTData);
        pCLUTData = NULL;
    }

    if (nRet != MV_OK)
    {
        if (handle != NULL)
        {
            MV_CC_DestroyHandle(handle);
            handle = NULL;
        }
    }

    printf("Press a key to exit.\n");
    WaitForKeyPress();

    return 0;
}
